README.md

```diff
@@ -1,34 +1,134 @@
-# Crate Name
+# event-manager
+
+The `event-manager` provides abstractions for implementing event based
+systems. This crate only works on Linux and uses the
+[epoll](http://man7.org/linux/man-pages/man7/epoll.7.html) API to provide a
+mechanism for handling I/O notifications.
 
 ## Design
 
-TODO: This section should have a high-level design of the crate.
+This crate is build around two abstractions:
+- Event Manager
+- Event Subscriber
+
+The Event Manager allows adding and removing subscribers, and provides
+APIs through which the subscribers can be updated in terms of events in the
+interest list. These actions are abstracted through the `SubscriberOps` trait.
+
+To interface with the Event Manager, the Event Subscribers need to provide an
+initialization function, and a callback for when events in the
+interest list become ready. The subscribers can update their interest list
+when handling ready events. These actions are abstracted through the
+`EventSubscriber` trait.
+
+A typical event-based application, creates the event manager, registers its
+subscribers, and then calls into the event manager's `run` function in a loop.
+Behind the scenes, the event manager calls into `epoll::wait` and maps the file
+descriptors in the ready list to the subscribers it manages. The event manager
+calls the subscriber's process function (its registered callback). When
+dispatching the events, the event manager creates a specialized object and
+passes it in the callback function so that the subscribers can use it to alter
+their interest list.
```

Comment:
A diagram for this whole paragraph would be fantastic

Reply:
I'll try with a diagram. I already added an example in the next commit from this PR. There will be even more examples added as part of #5 so I wouldn't fix this at the "code level" as well.

Reply:
Added a diagram, but I am not very good at that. Any suggestions into how to make it better are more than welcome.

Reply:
Diagram LGTM